doc/guides/prog_guide/cnet.rst

```diff
@@ -0,0 +1,400 @@
+^^^^^^^^^^
+
+Each CNET instance (one per defined thread) is created with a set of nodes with each node handling
+a number of packets at a time. Packets are processed from the inputs to the output nodes in
+the :ref:`CNET Graph Nodes <figure_cnet>` in a single thread.
+
+Input nodes like *eth_rx-N*, *kernel_recv* are called *source* nodes and are called each
+time the graph is walked to get more packets to process. The *udp_output* and *tcp_output* nodes
+are a special type of *source* node and are only called when the graph is walked when packets are
+added to the nodes. The CNET stack uses these nodes to place out bound packet data from the
+application via the *channel* APIs.
+
+  **Note**: The *chnl_callback* node calls back into the application using the same thread as the
+  CNET graph instance. Processing of packets needs to be done within the CNET stack thread to avoid
+  locks and race conditions.
+
+Input Nodes
+^^^^^^^^^^^
+
+- **eth_rx-N** is the Ethernet input node and is source node, which means it is called repeatedly.
+
+  - **Note**: More then one eth_rx node can be present one per lport.
+
+- **eth_tx-N** is the Ethernet ouput node and is called anytime packets are added to the node.
+
+  - **Note**: More then one eth_tx node can be present one per lport.
+
+- **kernel_recv** is another source node and is called to receive packets from the kernel.
+- **ptype** is the node to determine the packet type i.e. UDP/IPv4, TCP/IPv4, ... and the next node to call.
+- **gtpu_input** is the node to support GTPU packets (**WiP**)
+- **ip4_input** is the IPv4 input node for processing IPv4 packets, IPv6 node will be at this same level.
+- **ip4_forward** is the IPv4 forwarding node for packets that have been received and can be quickly.
+- **ip4_proto** is the node to determine the next node for L4 protocols i.e. UDP or TCP.
+- **tcp_input** is the starting node to process TCP packets, which each packet is processed in the *cnet_tcp_input* function.
+
+- **udp_input** is the starting node to process UDP packet, which each packet determined if it is to be processed by the graph instance.
+
+- **chnl_recv** is the node to help send packet data processed by UDP or TCP to the application via the *chnl_callback* node.
+
+- **punt_kernel** is the node to send packets data to the kernel to be processed if the packet is not being processed by the application.
+
+  - Sending these packets to the kernel allows other processes waiting for packets on standard linux sockets to process the packets.
+
+- **eth_tx-N** is the node to send packets out a lport via the *Pktdev* APIs.
+
+  - **Note**: More then one output node may be present one per lport.
+
+- **arp_request** is the node to send request the kernel stack send and process ARP packets needing to be sent by CNET.
+
+  - ARP packet processing is handled by the Linux kernel and not by CNET. When the CNET stack needs a destination MAC address and it is not known a ARP request is sent by the kernel.
```

Comment:
nit: suggest some commas: When teh CNET stack needs a destination MAC address, and it is not known, a ARP request"